### DISCUSSION TITLE
check(), capture_default_str() and default_str() not working?

### DISCUSSION THREAD

**Discussion by patrikhuber** [CLIUtils/CLI11#989] [Q&A]
Hi,

I'm a new (potential) CLI11 user, and I struggle quite a bit with the readme / docs.

I have an app with a couple of command-line arguments that are paths to files. These paths should have default values (e.g. `"../example/myfile.bin"`), but be overrideable with a command-line argument like `myapp.exe --file c:/folder/another_file.bin`. I'd also like `myapp.exe --help` to show the default value.

I've tried a few things:

```
int main(int argc, char *argv[])
{
    CLI::App app{"MyApp"};
    std::filesystem::path binpath1 = "../example/myfile.bin"
    std::filesystem::path binpath2 = "../example/myfile.bin"
    std::filesystem::path binpath3 = "../example/myfile.bin"

    app.add_option("-f,--file1", binpath1, "Path to a .bin file"); // this works but the path "../example/myfile.bin" won't show up in --help. I need it to show somewhere that this is a required argument but it's populated with a default value - running the app.exe without the arg will work and use the default value, but the user can overwrite it.

    app.add_option("-g,--file2", binpath2, "Path to a .bin file")->capture_default_str(); // this doesn't seem to change anything, and I don't understand the documentation, it's not clear to me what it says.

    app.add_option("-h,--file3", binpath3, "Path to a .bin file")->default_str(); // this doesn't seem to change anything either, and I also don't understand the documentation, it's not clear to me what it should do or change.

    try
    {
        app.parse(argc, argv);
    } catch (const CLI::ParseError& e)
    {
        return app.exit(e);
    }
```

I've also been trying to use `app.add_option("-f,--file", binpath, "Path to a .bin file")->check(CLI::ExistingFile);` but if I give it a non-existing file, it just happily continues - no error or warning, like I would expect. What am I missing?

I've also been trying to use `app.add_option("-f,--file", binpath, "Path to a .bin file")->option_text("<file>")`. This works and changes "TEXT" in the `--help` screen to "<file>". However, when I start my app with the argument missing, e.g. `myapp.exe -f`, then it outputs: `--file: 1 required TEXT missing Run with --help for more information.`, and it hasn't replaced "TEXT" with "<file>". Is this a bug or am I doing something wrong?

Thanks a lot!

(If it's relevant, I'm on Windows 11, latest VS / MSVC, CLI11 from vcpkg, using CMake, and /std:c++23.)

**Comment by phlptp** [comment]
here is my main

```C++
int main(int argc, char *argv[])
{
    CLI::App app{"MyApp"};
    std::filesystem::path binpath1 = "../example/myfile.bin";
        std::filesystem::path binpath2 = "../example/myfile.bin";
        std::filesystem::path binpath3 = "../example/myfile.bin";

        app.add_option("-f,--file1", binpath1, "Path to a .bin file"); // this works but the path "../example/myfile.bin" won't show up in --help. I need it to show somewhere that this is a required argument but it's populated with a default value - running the app.exe without the arg will work and use the default value, but the user can overwrite it.

    app.add_option("-g,--file2", binpath2, "Path to a .bin file")->capture_default_str(); // this doesn't seem to change anything, and I don't understand the documentation, it's not clear to me what it says.

    app.add_option("-i,--file3", binpath3, "Path to a .bin file")->default_str("defFile.bin"); // this doesn't seem to change anything either, and I also don't understand the documentation, it's not clear to me what it should do or change.

    try
    {
        app.parse(argc, argv);
    } catch (const CLI::ParseError& e)
    {
        return app.exit(e);
    }
}
```

Here is the output for help
```
MyApp
Usage: C:\CodeProjects\CLI11\build\examples\Debug\tester.exe [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  -f,--file1 TEXT             Path to a .bin file
  -g,--file2 TEXT ["../example/myfile.bin"]
                              Path to a .bin file
  -i,--file3 TEXT [defFile.bin]
                              Path to a .bin file
```

so I think the capture_default_str   which pulls the default string from the variable, and default_str which allows you to set it directly is working as expected.

**Reply by patrikhuber** [reply]
Hi @phlptp,

Unfortunately neither of these work for me - exactly the same code, different `--help` output. Could it be that this is a bug that was fixed since 2.3.2 came out? I can see 2.3.2 is over a year old (released Jan 2023). Unfortunately I am unable to try master, I rely on vcpkg for dependencies.

**Reply by phlptp** [reply]
I don't know it is possible some things have changed, but hopefully a 2.4 release is coming out very soon.

**Reply by patrikhuber** [reply]
A colleague of mine just tested my commit on his Linux (Ubuntu) machine, and unfortunately this simple code doesn't work for him:

```
    std::filesystem::path bin_path;
    app.add_option("-b,--bin", bin_path, "Path to .bin file")
        ->option_text("<file>")
        ->default_str("../data/file.bin");
```
If I run it on my Windows machine without argument, i.e. `myapp.exe`, it fills `bin_path` with `"../data/file.bin"`, but on his Linux machine, `bin_path` is just empty. It looks like we're either both doing something wrong (either fundamentally or something silly), or 2.3.2 on vcpkg is broken in more than one way, on more than one platform.

I'm ready to throw in the towel and move back to cxxopts (which has always worked great for me and everyone using my apps in the past, but CLI11 has a few extra features that tempted me). But I just saw that you guys did actually release 2.4 this morning. Your vcpkg port for 2.3.2 looks easy without any patches needed. I'm not very familiar with updating ports in vcpkg but I'll give it a quick go and see whether I can get it to work.

**Comment by phlptp** [comment]
If I set the third option to 
```C++
app.add_option("-i,--file3", binpath3, "Path to a .bin file")->default_str("defFile.bin")->check(CLI::ExistingFile);
```

now when I run with -i not_a_file.bin  I get
```sh
--file3: File does not exist: not_a_file.bin
Run with --help for more information.
```

the check only checks when an option is given,  it does not check the variable exists unless `run_callback_for_default` is specified then it will check the default when given and update the variable in this case.

**Reply by patrikhuber** [reply]
Hi @phlptp, thanks for the quick reply!

You're right, `->check(CLI::ExistingFile)` actually works if you pass it an invalid file (e.g. `-i not_a_file.bin`), and as you said (as expected), it doesn't run on default arguments.

However, it still doesn't check the default value if I do:
```
app.add_option("-i,--file3", binpath3, "Path to a .bin file")->default_str("defFile.bin")->check(CLI::ExistingFile)->run_callback_for_default();
```
`app.parse` runs happily, even though `defFile.bin` does not exist.

Could it be that this is a bug that was fixed since 2.3.2 came out? I can see 2.3.2 is over a year old (released Jan 2023). Unfortunately I am unable to try master, I rely on vcpkg for dependencies.

**Reply by phlptp** [reply]
```C++
app.add_option("-i,--file3", binpath3, "Path to a .bin file")->check(CLI::ExistingFile)->run_callback_for_default()->default_str("defFile.bin");
```

It would likely need to be in this order for that to work.  You would have change the setting on running the callback and add the validator before setting the default.

**Reply by patrikhuber** [reply]
> It would likely need to be in this order for that to work. You would have change the setting on running the callback and add the validator before setting the default.

That doesn't work either unfortunately.

**Comment by henryiii** [comment]
`capture_default_str()` converts the given argument into a string and uses that for the `default_str(value)` (which you can set manually, as well). Not all types support being converted to strings, and you might not have initialized a value even if it does support conversion to strings.

If those are now showing up, that sounds like a bug. Does it work with just plain std::string? We do have a few tests for std::filesystem, but not a lot, and possibly it behaves differently on Windows.

**Reply by patrikhuber** [reply]
I just changed the `std::filesystem::path`'s to `std::string`, and to be honest, nothing seems to have changed. The default paths are still not showing up in the help, the default files are still not being checked for existence, and `capture_default_str()` also doesn't capture the value and show it in the help like it does for Philip. It looks to me like 2.3.2 is crooked, or perhaps for some weird reason the vcpkg build. I really hope I'm not making any stupid rookie mistakes but I can't imagine what that would be.
I can test v2.4.0 once it's merged into vcpkg.

**Reply by henryiii** [reply]
Okay, I'm also adding some tests. Just to point out above, `->default_string(...)` takes the value of the default string, with no arguments removing the default string entirely. So it's expected that `x->default_string()` (`--file3` example) would produce no default string.

**Reply by patrikhuber** [reply]
> it's expected that x->default_string() (--file3 example) would produce no default string.

That makes sense and is the behaviour I'd expect. It isn't what's causing my issues. But good to mention it!
